test/unit/active_record_batch_enumerator_test.rb

```diff
@@ -0,0 +1,103 @@
+        assert batch.is_a?(ActiveRecord::Relation)
+        assert_equal product_batches[index].first, batch
+        assert_equal product_batches[index].last, cursor
+      end
+    end
+
+    test "#each yields unloaded relations" do
+      enum = build_enumerator
+      relation, _ = enum.first
+
+      refute_predicate relation, :loaded?
+    end
+
+    test "#each doesn't yield anything if the relation is empty" do
+      enum = build_enumerator(relation: Product.none)
+
+      assert_equal([], enum.to_a)
+    end
+
+    test "batch size is configurable" do
+      enum = build_enumerator(batch_size: 4)
+      products = Product.order(:id).take(4)
+
+      assert_equal([products, products.last.id], enum.first)
+    end
+
+    test "columns are configurable" do
+      enum = build_enumerator(columns: [:updated_at])
+      products = Product.order(:updated_at).take(2)
+
+      expected_product_cursor = products.last.updated_at.strftime(SQL_TIME_FORMAT)
+      assert_equal([products, expected_product_cursor], enum.first)
+    end
+
+    test "columns can be an array" do
+      enum = build_enumerator(columns: [:updated_at, :id])
+      products = Product.order(:updated_at, :id).take(2)
+
+      expected_product_cursor = [products.last.updated_at.strftime(SQL_TIME_FORMAT), products.last.id]
+      assert_equal([products, expected_product_cursor], enum.first)
+    end
+
+    test "columns configured with primary key only queries primary key column once" do
+      queries = track_uncached_queries do
+        enum = build_enumerator(columns: [:updated_at, :id])
+        enum.first
+      end
+      assert_equal 1, queries.first.scan(/`products`.`id`/).count
+    end
+
+    test "cursor can be used to resume" do
+      products = Product.order(:id).take(3)
+
+      enum = build_enumerator(cursor: products.shift.id)
+
+      assert_equal([products, products.last.id], enum.first)
+    end
+
+    test "cursor can be used to resume on multiple columns" do
+      enum = build_enumerator(columns: [:created_at, :id])
+      products = Product.order(:created_at, :id).take(2)
+
+      cursor = [products.last.created_at.strftime(SQL_TIME_FORMAT), products.last.id]
+      assert_equal([products, cursor], enum.first)
+
+      enum = build_enumerator(columns: [:created_at, :id], cursor: cursor)
+      products = Product.order(:created_at, :id).offset(2).take(2)
+
+      cursor = [products.last.created_at.strftime(SQL_TIME_FORMAT), products.last.id]
+      assert_equal([products, cursor], enum.first)
+    end
+
+    private
+
+    def build_enumerator(relation: Product.all, batch_size: 2, columns: nil, cursor: nil)
+      JobIteration::ActiveRecordBatchEnumerator.new(
+        relation,
+        batch_size: batch_size,
+        columns: columns,
+        cursor: cursor,
+      )
+    end
+
+    def track_uncached_queries(&block)
+      queries = []
+      query_cb = ->(*, payload) { queries << payload[:sql] unless payload[:cached] }
```

Comment:
I think we don't need the cached/uncached distinction anymore, since we don't do cached queries anymore?